### ISSUE TITLE
FTL crashes when adding domain to audit

### ISSUE THREAD

**Issue by ics** [pi-hole/FTL#841]
**In raising this issue, I confirm the following (please check boxes, eg [X]) Failure to fill the template will close your issue:**

- [x] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md).
- [x] The issue I am reporting can be *replicated*
- [x] The issue I am reporting isn't a duplicate 

**How familiar are you with the codebase?:**

1

---

**[BUG | ISSUE] Expected Behaviour:**
Stable running pihole-FTL.

**[BUG | ISSUE] Actual Behaviour:**
FTL crashes when adding domain to audit.

**[BUG | ISSUE] Steps to reproduce:**

- Run Pi-hole v5.1.1 Web Interface v5.1 FTL v5.1
- Go to admin interface -> Tools -> Audit Log
- Click 'Audit' in either Allowed or Block queries for 10-15 domains
- FTL crashes

**Log file output [if available]**

```
[2020-07-18 03:45:29.152 17328/T18219] FTL has been running for 507 seconds
[2020-07-18 03:45:29.153 17328/T18219] FTL branch: master
[2020-07-18 03:45:29.153 17328/T18219] FTL version: v5.1
[2020-07-18 03:45:29.153 17328/T18219] FTL commit: b522646
[2020-07-18 03:45:29.153 17328/T18219] FTL date: 2020-07-15 22:26:33 +0100
[2020-07-18 03:45:29.154 17328/T18219] FTL user: started as pihole, ended as pihole
[2020-07-18 03:45:29.155 17328/T18219] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
[2020-07-18 03:45:29.155 17328/T18219] Process details: MID: 17328
[2020-07-18 03:45:29.155 17328/T18219]                  PID: 17328
[2020-07-18 03:45:29.155 17328/T18219]                  TID: 18219
[2020-07-18 03:45:29.156 17328/T18219]                  Name: telnet-21
[2020-07-18 03:45:29.156 17328/T18219] Received signal: Segmentation fault
[2020-07-18 03:45:29.156 17328/T18219]      at address: (nil)
[2020-07-18 03:45:29.156 17328/T18219]      with code:  Unknown (128)
[2020-07-18 03:45:29.163 17328/T18219] Backtrace:
[2020-07-18 03:45:29.163 17328/T18219] B[0000]: /usr/bin/pihole-FTL(+0x342af) [0x5607b9f502af]
[2020-07-18 03:45:29.169 17328/T18219] L[0000]: N/A (0x342af)
[2020-07-18 03:45:29.175 17328/T18219] B[0001]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x12730) [0x7f0d2b65f730]
[2020-07-18 03:45:29.175 17328/T18219] B[0002]: /lib/x86_64-linux-gnu/libpthread.so.0(__pthread_mutex_lock+0) [0x7f0d2b6576c0]
[2020-07-18 03:45:29.175 17328/T18219] B[0003]: /usr/bin/pihole-FTL(sqlite3_column_int+0x28) [0x5607b9fd2ce8]
[2020-07-18 03:45:29.181 17328/T18219] L[0003]: N/A (0xb6ce8)
[2020-07-18 03:45:29.187 17328/T18219] B[0004]: /usr/bin/pihole-FTL(in_auditlist+0x19a) [0x5607b9f5929a]
[2020-07-18 03:45:29.190 17328M] Compiled 0 whitelist and 62 blacklist regex filters for 62 clients in 56.0 msec
[2020-07-18 03:45:29.193 17328/T18219] L[0004]: N/A (0x3d29a)
[2020-07-18 03:45:29.199 17328/T18219] B[0005]: /usr/bin/pihole-FTL(getTopDomains+0x279) [0x5607b9f51519]
[2020-07-18 03:45:29.205 17328/T18219] L[0005]: N/A (0x35519)
[2020-07-18 03:45:29.207 17328/T18219] B[0006]: /usr/bin/pihole-FTL(process_request+0x1bd) [0x5607b9f543dd]
[2020-07-18 03:45:29.210 17328/T18219] L[0006]: N/A (0x383dd)
[2020-07-18 03:45:29.211 17328/T18219] B[0007]: /usr/bin/pihole-FTL(+0x38ad6) [0x5607b9f54ad6]
[2020-07-18 03:45:29.215 17328/T18219] L[0007]: N/A (0x38ad6)
[2020-07-18 03:45:29.215 17328/T18219] B[0008]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3) [0x7f0d2b654fa3]
[2020-07-18 03:45:29.215 17328/T18219] B[0009]: /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f0d2b5854cf]
[2020-07-18 03:45:29.215 17328/T18219] ------ Listing content of directory /dev/shm ------
[2020-07-18 03:45:29.215 17328/T18219] File Mode User:Group  Filesize Filename
[2020-07-18 03:45:29.216 17328/T18219] rwxrwxrwx root:root 260 .
[2020-07-18 03:45:29.216 17328/T18219] rwxr-xr-x root:root 3K ..
[2020-07-18 03:45:29.216 17328/T18219] rw------- pihole:pihole 4K FTL-per-client-regex
[2020-07-18 03:45:29.216 17328/T18219] rw------- pihole:pihole 4K FTL-dns-cache
[2020-07-18 03:45:29.216 17328/T18219] rw------- pihole:pihole 45K FTL-overTime
[2020-07-18 03:45:29.216 17328/T18219] rw------- pihole:pihole 4M FTL-queries
[2020-07-18 03:45:29.217 17328/T18219] rw------- pihole:pihole 4K FTL-upstreams
[2020-07-18 03:45:29.217 17328/T18219] rw------- pihole:pihole 332K FTL-clients
[2020-07-18 03:45:29.217 17328/T18219] rw------- pihole:pihole 98K FTL-domains
[2020-07-18 03:45:29.217 17328/T18219] rw------- pihole:pihole 102K FTL-strings
[2020-07-18 03:45:29.217 17328/T18219] rw------- pihole:pihole 12 FTL-settings
[2020-07-18 03:45:29.217 17328/T18219] rw------- pihole:pihole 144 FTL-counters
[2020-07-18 03:45:29.217 17328/T18219] rw------- pihole:pihole 48 FTL-lock
```

```
[2020-07-18 04:11:39.276 491/T1657] FTL has been running for 1034 seconds
[2020-07-18 04:11:39.276 491/T1657] FTL branch: master
[2020-07-18 04:11:39.276 491/T1657] FTL version: v5.1
[2020-07-18 04:11:39.276 491/T1657] FTL commit: b522646
[2020-07-18 04:11:39.276 491/T1657] FTL date: 2020-07-15 22:26:33 +0100
[2020-07-18 04:11:39.277 491/T1657] FTL user: started as pihole, ended as pihole
[2020-07-18 04:11:39.277 491/T1657] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
[2020-07-18 04:11:39.277 491/T1657] Process details: MID: 491
[2020-07-18 04:11:39.277 491/T1657]                  PID: 491
[2020-07-18 04:11:39.277 491/T1657]                  TID: 1657
[2020-07-18 04:11:39.277 491/T1657]                  Name: telnet-21
[2020-07-18 04:11:39.278 491/T1657] Received signal: Segmentation fault
[2020-07-18 04:11:39.278 491/T1657]      at address: 0x7f16d8768997
[2020-07-18 04:11:39.278 491/T1657]      with code:  SEGV_MAPERR (Address not mapped to object)
[2020-07-18 04:11:39.281 491/T1657] Backtrace:
[2020-07-18 04:11:39.281 491/T1657] B[0000]: /usr/bin/pihole-FTL(+0x342af) [0x55edac4f52af]
[2020-07-18 04:11:39.284 491/T1657] L[0000]: N/A (0x342af)
[2020-07-18 04:11:39.285 491/T1657] B[0001]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x12730) [0x7f16d89ca730]
[2020-07-18 04:11:39.285 491/T1657] B[0002]: /lib/x86_64-linux-gnu/libc.so.6(+0x15f194) [0x7f16d8956194]
[2020-07-18 04:11:39.285 491/T1657] B[0003]: /usr/bin/pihole-FTL(+0x952b4) [0x55edac5562b4]
[2020-07-18 04:11:39.288 491/T1657] L[0003]: N/A (0x952b4)
[2020-07-18 04:11:39.288 491/T1657] B[0004]: /usr/bin/pihole-FTL(+0xe052a) [0x55edac5a152a]
[2020-07-18 04:11:39.290 491/T1657] L[0004]: N/A (0xe052a)
[2020-07-18 04:11:39.291 491/T1657] B[0005]: /usr/bin/pihole-FTL(+0x12e578) [0x55edac5ef578]
[2020-07-18 04:11:39.294 491/T1657] L[0005]: N/A (0x12e578)
[2020-07-18 04:11:39.294 491/T1657] B[0006]: /usr/bin/pihole-FTL(+0x136868) [0x55edac5f7868]
[2020-07-18 04:11:39.297 491/T1657] L[0006]: N/A (0x136868)
[2020-07-18 04:11:39.297 491/T1657] B[0007]: /usr/bin/pihole-FTL(in_auditlist+0xb4) [0x55edac4fe1b4]
[2020-07-18 04:11:39.300 491/T1657] L[0007]: N/A (0x3d1b4)
[2020-07-18 04:11:39.300 491/T1657] B[0008]: /usr/bin/pihole-FTL(getTopDomains+0x279) [0x55edac4f6519]
[2020-07-18 04:11:39.303 491/T1657] L[0008]: N/A (0x35519)
[2020-07-18 04:11:39.303 491/T1657] B[0009]: /usr/bin/pihole-FTL(process_request+0x1bd) [0x55edac4f93dd]
[2020-07-18 04:11:39.306 491/T1657] L[0009]: N/A (0x383dd)
[2020-07-18 04:11:39.306 491/T1657] B[0010]: /usr/bin/pihole-FTL(+0x38ad6) [0x55edac4f9ad6]
[2020-07-18 04:11:39.309 491/T1657] L[0010]: N/A (0x38ad6)
[2020-07-18 04:11:39.309 491/T1657] B[0011]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3) [0x7f16d89bffa3]
[2020-07-18 04:11:39.309 491/T1657] B[0012]: /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f16d88f04cf]
[2020-07-18 04:11:39.309 491/T1657] ------ Listing content of directory /dev/shm ------
[2020-07-18 04:11:39.309 491/T1657] File Mode User:Group  Filesize Filename
[2020-07-18 04:11:39.310 491/T1657] rwxrwxrwx root:root 260 .
[2020-07-18 04:11:39.310 491/T1657] rwxr-xr-x root:root 3K ..
[2020-07-18 04:11:39.310 491/T1657] rw------- pihole:pihole 4K FTL-per-client-regex
[2020-07-18 04:11:39.310 491/T1657] rw------- pihole:pihole 8K FTL-dns-cache
[2020-07-18 04:11:39.310 491/T1657] rw------- pihole:pihole 45K FTL-overTime
[2020-07-18 04:11:39.310 491/T1657] rw------- pihole:pihole 4M FTL-queries
[2020-07-18 04:11:39.310 491/T1657] rw------- pihole:pihole 4K FTL-upstreams
[2020-07-18 04:11:39.310 491/T1657] rw------- pihole:pihole 332K FTL-clients
[2020-07-18 04:11:39.311 491/T1657] rw------- pihole:pihole 98K FTL-domains
[2020-07-18 04:11:39.311 491/T1657] rw------- pihole:pihole 106K FTL-strings
[2020-07-18 04:11:39.311 491/T1657] rw------- pihole:pihole 12 FTL-settings
[2020-07-18 04:11:39.311 491/T1657] rw------- pihole:pihole 144 FTL-counters
[2020-07-18 04:11:39.311 491/T1657] rw------- pihole:pihole 48 FTL-lock
```
```
[2020-07-18 04:29:51.165 1751/T2293] FTL has been running for 1052 seconds
[2020-07-18 04:29:51.166 1751/T2293] FTL branch: master
[2020-07-18 04:29:51.166 1751/T2293] FTL version: v5.1
[2020-07-18 04:29:51.166 1751/T2293] FTL commit: b522646
[2020-07-18 04:29:51.167 1751/T2293] FTL date: 2020-07-15 22:26:33 +0100
[2020-07-18 04:29:51.167 1751/T2293] FTL user: started as pihole, ended as pihole
[2020-07-18 04:29:51.168 1751/T2293] Compiled for x86_64 (compiled on CI) using gcc (Debian 6.3.0-18+deb9u1) 6.3.0 20170516
[2020-07-18 04:29:51.168 1751/T2293] Process details: MID: 1751
[2020-07-18 04:29:51.168 1751/T2293]                  PID: 1751
[2020-07-18 04:29:51.169 1751/T2293]                  TID: 2293
[2020-07-18 04:29:51.169 1751/T2293]                  Name: telnet-23
[2020-07-18 04:29:51.169 1751/T2293] Received signal: Segmentation fault
[2020-07-18 04:29:51.169 1751/T2293]      at address: (nil)
[2020-07-18 04:29:51.170 1751/T2293]      with code:  Unknown (128)
[2020-07-18 04:29:51.171 1751/T2293] Backtrace:
[2020-07-18 04:29:51.171 1751/T2293] B[0000]: /usr/bin/pihole-FTL(+0x342af) [0x5570533062af]
[2020-07-18 04:29:51.177 1751/T2293] L[0000]: N/A (0x342af)
[2020-07-18 04:29:51.179 1751/T2293] B[0001]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x12730) [0x7f0e8cba8730]
[2020-07-18 04:29:51.179 1751/T2293] B[0002]: /lib/x86_64-linux-gnu/libpthread.so.0(__pthread_mutex_lock+0) [0x7f0e8cba06c0]
[2020-07-18 04:29:51.180 1751/T2293] B[0003]: /usr/bin/pihole-FTL(sqlite3_clear_bindings+0x23) [0x557053388523]
[2020-07-18 04:29:51.185 1751/T2293] L[0003]: N/A (0xb6523)
[2020-07-18 04:29:51.191 1751/T2293] B[0004]: /usr/bin/pihole-FTL(in_auditlist+0x1b6) [0x55705330f2b6]
[2020-07-18 04:29:51.197 1751/T2293] L[0004]: N/A (0x3d2b6)
[2020-07-18 04:29:51.199 1751M] Compiled 0 whitelist and 61 blacklist regex filters for 64 clients in 55.5 msec
[2020-07-18 04:29:51.199 1751/T2293] B[0005]: /usr/bin/pihole-FTL(getTopDomains+0x279) [0x557053307519]
[2020-07-18 04:29:51.205 1751/T2293] L[0005]: N/A (0x35519)
[2020-07-18 04:29:51.207 1751/T2293] B[0006]: /usr/bin/pihole-FTL(process_request+0x1bd) [0x55705330a3dd]
[2020-07-18 04:29:51.213 1751/T2293] L[0006]: N/A (0x383dd)
[2020-07-18 04:29:51.215 1751/T2293] B[0007]: /usr/bin/pihole-FTL(+0x38ad6) [0x55705330aad6]
[2020-07-18 04:29:51.220 1751/T2293] L[0007]: N/A (0x38ad6)
[2020-07-18 04:29:51.221 1751/T2293] B[0008]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x7fa3) [0x7f0e8cb9dfa3]
[2020-07-18 04:29:51.221 1751/T2293] B[0009]: /lib/x86_64-linux-gnu/libc.so.6(clone+0x3f) [0x7f0e8cace4cf]
[2020-07-18 04:29:51.221 1751/T2293] ------ Listing content of directory /dev/shm ------
[2020-07-18 04:29:51.221 1751/T2293] File Mode User:Group  Filesize Filename
[2020-07-18 04:29:51.221 1751/T2293] rwxrwxrwx root:root 260 .
[2020-07-18 04:29:51.221 1751/T2293] rwxr-xr-x root:root 3K ..
[2020-07-18 04:29:51.222 1751/T2293] rw------- pihole:pihole 4K FTL-per-client-regex
[2020-07-18 04:29:51.222 1751/T2293] rw------- pihole:pihole 4K FTL-dns-cache
[2020-07-18 04:29:51.222 1751/T2293] rw------- pihole:pihole 45K FTL-overTime
[2020-07-18 04:29:51.222 1751/T2293] rw------- pihole:pihole 4M FTL-queries
[2020-07-18 04:29:51.222 1751/T2293] rw------- pihole:pihole 4K FTL-upstreams
[2020-07-18 04:29:51.222 1751/T2293] rw------- pihole:pihole 332K FTL-clients
[2020-07-18 04:29:51.222 1751/T2293] rw------- pihole:pihole 98K FTL-domains
[2020-07-18 04:29:51.223 1751/T2293] rw------- pihole:pihole 102K FTL-strings
[2020-07-18 04:29:51.223 1751/T2293] rw------- pihole:pihole 12 FTL-settings
[2020-07-18 04:29:51.223 1751/T2293] rw------- pihole:pihole 144 FTL-counters
[2020-07-18 04:29:51.223 1751/T2293] rw------- pihole:pihole 48 FTL-lock
```

**Device specifics**

Hardware Type: bhyve virtual machine
OS: Debian GNU/Linux 10 (buster) - 4.19.0-9-amd64

**Comment by DL6ER** [comment]
@ics Please run
```
addr2line 0x342af 0xb6523 0x3d2b6 0x35519 0x383dd 0x38ad6 -e /usr/bin/pihole-FTL
```
and post the output

**Comment by ics** [comment]
```
# addr2line 0x342af 0xb6523 0x3d2b6 0x35519 0x383dd 0x38ad6 -e /usr/bin/pihole-FTL
/root/project/src/signals.c:119
/root/project/src/database/sqlite3.c:82668 (discriminator 1)
/root/project/src/database/gravity-db.c:752
/root/project/src/api/api.c:304 (discriminator 1)
/root/project/src/api/request.c:52
/root/project/src/api/socket.c:315
```

**Comment by DL6ER** [comment]
Thanks I can reproduce this now and also already know *why* it is happening. It will just need some more thoughts into how to fix it exactly. It's a concurrency problem when we're reloading the gravity database connection and query the database at the same time.

The fix for this should be rather simple, however, it is something for tomorrow. It already got too late for today.

**Comment by ics** [comment]
Great! Glad I could help. Thank you for looking into it.

**Comment by DL6ER** [comment]
Sometimes you shouldn't stray up for doing this, but oh well...

Please try if
```
pihole checkout ftl fix/guard_gravity_reopening
```
fixes the issues you're seeing.

**Comment by ics** [comment]
Yes, FTL no longer crashes. However, the audit button adds the domain to whitelist/blacklist instead of auditing.
That seems to be a separate issue though.

**Comment by DL6ER** [comment]
Thanks, I will check out the web audit feature when I'm back from work today.

**Comment by pralor** [comment]
This issue has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/audit-log-dns-ftl-aussetzer-nach-update-auf-5-1-1/35623/24

**Comment by yubiuser** [comment]
@DL6ER 
>Thanks, I will check out the web audit feature when I'm back from work today.
https://github.com/pi-hole/AdminLTE/issues/1470

**Comment by DL6ER** [comment]
https://github.com/pi-hole/AdminLTE/issues/1470 should be fixed by https://github.com/pi-hole/AdminLTE/pull/1518

**Comment by ics** [comment]
> [pi-hole/AdminLTE#1470](https://github.com/pi-hole/AdminLTE/issues/1470) should be fixed by [pi-hole/AdminLTE#1518](https://github.com/pi-hole/AdminLTE/pull/1518)

Yes. Running stable now with:
```
  Pi-hole version is v5.1.1 (Latest: v5.1.1)
  AdminLTE version is fix/audit_log v5.0-464-gfdea341 (Latest: v5.1)
  FTL version is fix/guard_gravity_reopening vDev-3af4f87 (Latest: v5.1)
```

Thank you.

**Comment by DL6ER** [comment]
This should be fixed by the release of Pi-hole FTL 5.2. Feel free to re-open of create a new ticket with new information if the issue persists.
